typing/printtyp.ml

```diff
@@ -1229,8 +1229,17 @@ let rec tree_of_type_decl id decl =
     let vari =
       List.map2
         (fun ty v ->
-          if abstr || not (is_Tvar (repr ty)) then Variance.get_upper v
-          else (true,true))
+          let is_var = is_Tvar (repr ty) in
+          if abstr || not is_var then
+            let inj =
+              decl.type_kind = Type_abstract && Variance.mem Inj v &&
+              match decl.type_manifest with
+              | None -> true
+              | Some ty -> (* only abstract or private row types *)
+                  decl.type_private = Private &&
+                  Btype.is_constr_row ~allow_ident:true (Btype.row_of_type ty)
```

Comment:
I'm not sure about these heuristics for printing injectivity annotations, which sometimes lead to surprising behaviour.  For example, when a type definition is printed at the top level, sometimes injectivity annotations are dropped and variance annotations are retained:

```ocaml
# type +!'a t = 'a constraint 'a = 'b * int;;
type +'a t = 'a constraint 'a = 'b * int
```

Reply:
As explained, injectivity is inferred for all types except abstract types and private row types. When it is inferred, there is no need to print it.
Currently, variance inference is much weaker. It might be possible to make it stronger, to avoid such discrepancies on constrained type parameters, but this is a problem with variance, not injectivity.